File: python/1261.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class FindElements:

    def __init__(self, root: Optional[TreeNode]):
        self.root = root

    def find(self, target: int) -> bool:
        target += 1
        cur = self.root
        for i in range(target.bit_length() - 2, -1, -1):
            bit = (target >> i) & 1
            if bit == 0:
                cur = cur.left
            else:
                cur = cur.right
            if cur == None:
                return False
        return True

# class FindElements:

#     def __init__(self, root: Optional[TreeNode]):
#         s = set()
#         def dfs(node: Optional[TreeNode], val: int) -> None:
#             if node is None:
#                 return
#             s.add(val)
#             dfs(node.left, val * 2 + 1)
#             dfs(node.right, val * 2 + 2)
#         dfs(root, 0)
#         self.s = s

#     def find(self, target: int) -> bool:
#         return target in self.s



# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)

File: go/1261.go
package main

import "math/bits"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type FindElements struct {
	root *TreeNode
}

func Constructor(root *TreeNode) FindElements {
	return FindElements{root}
}

func (f *FindElements) Find(target int) bool {
	target++
	cur := f.root
	for i := bits.Len(uint(target)) - 2; i > -1; i-- {
		bit := target >> i & 1
		if bit == 0 {
			cur = cur.Left
		} else {
			cur = cur.Right
		}
		if cur == nil {
			return false
		}
	}
	return true
}

/**
 * Your FindElements object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Find(target);
 */

// type FindElements map[int]bool

// func Constructor(root *TreeNode) FindElements {
// 	f := FindElements{}
// 	var dfs func(node *TreeNode, val int)
// 	dfs = func(node *TreeNode, val int) {
// 		if node == nil {
// 			return
// 		}
// 		f[val] = true
// 		dfs(node.Left, val*2+1)
// 		dfs(node.Right, val*2+2)
// 	}
// 	dfs(root, 0)
// 	return f
// }

// func (f FindElements) Find(target int) bool {
// 	return f[target]
// }

/**
 * Your FindElements object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Find(target);
 */
